osm-fudge/bk_tree.py

```diff
@@ -0,0 +1,37 @@
+'''An implementation of bk-tree
+
+More: https://en.wikipedia.org/wiki/BK-tree
+'''
+
+import levenshtein
+
+class BKTree:
+    '''A bk-tree implementation and defined user methods
+       for insertion and querying the string
+    '''
+
+    def __init__(self, root):
+        self.tree = {root: {}}
+
+    def insert(self, str1, subtree):
```

Comment:
I think the public interface here should hide the subtree argument.

Maybe make a private helper method `_insert` and in your public method here call that?

Reply:
I changed the function definitions as following:

```
    def insert(self, str1, subtree):
        self.__insert(str1, subtree)

    def __insert(self, str1, subtree):
        root = list(subtree.keys())[0]
        dist = self.distanceMetrics(root, str1)

        if (dist in subtree[root]):
            self.insert(str1, subtree[root][dist])
        else:
            subtree[root][dist] = {str1: {}}
```

But still didn't understand what does hiding `subtree` in insert method means?

Reply:
The publicly facing API should not have the subtree argument in it.

All users care about is:
- they want to construct a bktree object by passing in their distance metric
- now with the bktree object they want to insert an an item (doesn't even have to be a string :P)

Users should not have to know about the implementation details with the subtree argument. Instead the public APU should be e.g. `def insert(self, item)`.

Reply:
Got you!! Abstraction 👍